File: 슬라이딩 윈도우/leetcode904(Medium)-Fruit Into Baskets.js
// 해시 방식의 슬라이딩 윈도우 : O(N), 460ms(21.26%), 57.8MB(28.68%)
/**
 * @param {number[]} fruits
 * @return {number}
 */
var totalFruit = function (fruits) {
  let result = 0
  let left = 0
  const baskets = {}

  for (let right = 0; right < fruits.length; right += 1) {
    const fruit = fruits[right]
    baskets[fruit] = (baskets[fruit] || 0) + 1

    while (Object.keys(baskets).length > 2) {
      baskets[fruits[left]] -= 1
      if (baskets[fruits[left]] === 0) {
        delete baskets[fruits[left]]
      }
      left += 1
    }

    result = Math.max(result, right - left + 1)
  }

  return result
}

File: 슬라이딩 윈도우/leetcode904(Medium)-Fruit Into Baskets.py
# 해시 방식의 슬라이딩 윈도우 : O(N), 973ms(70.47%), 20.1MB(80.52%)
class Solution:
    def totalFruit(self, fruits: List[int]) -> int:
        result = 0
        left = 0
        baskets = {}
        
        for right, fruit in enumerate(fruits):
            baskets[fruit] = baskets.get(fruit, 0) + 1

            while len(baskets) > 2:
                baskets[fruits[left]] -= 1
                if baskets[fruits[left]] == 0:
                    del baskets[fruits[left]]
                left += 1

            result = max(result, right - left + 1)
        
        return result